loki/visitors/pretty_visualize.py

```diff
@@ -0,0 +1,349 @@
+# (C) Copyright 2018- ECMWF.
+# This software is licensed under the terms of the Apache Licence Version 2.0
+# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+# In applying this licence, ECMWF does not waive the privileges and immunities
+# granted to it by virtue of its status as an intergovernmental organisation
+# nor does it submit to any jurisdiction.
+
+"""
+Pretty-Graph-Visualizer classes for IR
+"""
+from itertools import chain
+from codetiming import Timer
+from graphviz import Digraph
+from loki.tools import JoinableStringList, is_iterable, as_tuple
+from loki.visitors.visitor import Visitor
+
+__all__ = ["Visualizer", "pretty_visualize"]
+
+
+class Visualizer(Visitor):
+    """
+    Convert a given IR tree to a node and edge list via the visit mechanism.
+
+    This serves as base class for backends and provides a number of helpful
+    routines that ease implementing automatic recursion and line wrapping.
+    It is adapted from the Stringifier in "pprint.py". It doubles as a means
+    to produce a human readable graph representation of the IR, which is
+    useful for debugging purposes and first visualization.
+
+    Parameters
+    ----------
+    linewidth : int, optional
+        The line width limit after which to break a line.
+    symgen : str, optional
+        A function handle that accepts a :any:`pymbolic.primitives.Expression`
+        and produces a string representation for that.
+    """
+
+    def __init__(
+        self,
+        linewidth=40,
+        symgen=str,
+    ):
+        super().__init__()
+        self.linewidth = linewidth
+        self._symgen = symgen
+        self._id = 0
+        self._id_map = {}
+
+    @property
+    def symgen(self):
+        """
+        Formatter for expressions.
+        """
+        return self._symgen
+
+    def join_items(self, items, sep=", ", separable=True):
+        """
+        Concatenate a list of items into :any:`JoinableStringList`.
+
+        The return value can be passed to :meth:`format_line` or
+        :meth:`format_node` or converted to a string with `str`, using
+        the :any:`JoinableStringList` as an argument.
+        Upon expansion, lines will be wrapped automatically to stay within
+        the linewidth limit.
+
+        Parameters
+        ----------
+        items : list
+            The list of strings to be joined.
+        sep : str, optional
+            The separator to be inserted between items.
+        separable : bool, optional
+            Allow line breaks between individual :data:`items`.
+
+        Returns
+        -------
+        :any:`JoinableStringList`
+        """
+        return JoinableStringList(
+            items,
+            sep=sep,
+            width=self.linewidth,
+            cont="\n",
+            separable=separable,
+        )
+
+    def format_node(self, name, *items):
```

Comment:
[no action required] The duplication here compared to Stringifier makes me think that we should introduce a common base class for the joint functionality.